patches/@types/graphql+0.13.3.patch

```diff
@@ -0,0 +1,15 @@
+patch-package
+--- a/node_modules/@types/graphql/type/definition.d.ts
++++ b/node_modules/@types/graphql/type/definition.d.ts
+@@ -371,7 +371,10 @@ export interface GraphQLResolveInfo {
+     readonly path: ResponsePath;
+     readonly schema: GraphQLSchema;
+     readonly fragments: { [key: string]: FragmentDefinitionNode };
+-    readonly rootValue: any;
++    // Check out ./src/types/graphql.d.ts for the replacement
++    // which uses our types instead.
++    //
++    // readonly rootValue: any;
+     readonly operation: OperationDefinitionNode;
+     readonly variableValues: { [variableName: string]: any };
+ }
```

Comment:
Step 1: Remove the `readonly rootValue: any` from the GraphQL types. This means that we're free to overwrite it in some upcoming code.

This is odd, and conceptually shouldn't be needed because typescript interfaces are re-openable. e..g

```ts
interface MyThing {
  a: string
}

interface MyThing {
  b: number
}
```

is the same as:

```ts
interface MyThing {
  a: string
  b: number
}
```

Which you can see in action when I'm extending Jest's `expect` [object in a library](https://github.com/jest-community/jest-snapshots-svg/blob/b647082cc5b67bb21a187749f0840755a840ca54/src/index.ts#L6-L16)

```ts
declare global {
    namespace jest {
        interface Matchers<R> {
            /** Checks and sets up SVG rendering for React Components. */
            toMatchSVGSnapshot(width: number, height: number, settings?: { wireframe?: boolean }): void;
        }
    }
}
```

Which means if I should be able to do:

```ts
declare module "graphql" {
  interface GraphQLResolveInfo {
    /**
     * Where we store user/data-loader context, yes there's a context
     * object but without a big refactor it's not happening soon.
     */
    readonly rootValue: RootValue
  }
}

```

Which would mean any reference to `rootValue` in a resolver should look like this:

```ts
// theirs
export interface GraphQLResolveInfo {
    readonly rootValue: any
}

// ours
export interface GraphQLResolveInfo {
    readonly rootValue: RootValue
}

```

But, that doesn't work, see [this playground](https://www.typescriptlang.org/play/#src=interface%20GraphQLResolveInfo%20%7B%0D%0A%20%20readonly%20rootValue%3A%20any%0D%0A%7D%0D%0A%0D%0Ainterface%20GraphQLResolveInfo%20%7B%0D%0A%20%20readonly%20rootValue%3A%20%7B%20a%3A%20string%7D%0D%0A%7D%0D%0A%0D%0Aconst%20a%3A%20GraphQLResolveInfo%20%3D%20%7B%7D%20as%20any%0D%0Aa.rootValue)

So, instead I have patch-package remove the original definition and I have it get filled by our `graphql.d.ts`.

Reply:
Afaik you can only _add_ when merging declarations, only when extending can you make property types more specific. Seeing as we still need to explicitly type all our exported fields anyways, it should be fine to define our own extended interface instead of monkey-patching the default one.

Reply:
Hmm, this ended up being a bit annoying to do when using `GraphQLObjectType` as we're not creating a subclass and overriding `fields` on `GraphQLObjectTypeConfig` isn't possible due to the `Thunk<…>`. I think we can still improve this and also deal with not having to pass in `TContext` everywhere, but we can iterate on that.

Reply:
In reality, this might be the reason for us to move to using the proper ‘context’ arg to put the loaders in, as the context is already typeable. /cc @damassi

Reply:
Seems worth it